### DISCUSSION TITLE
deck.gl in anywidget

### DISCUSSION THREAD

**Discussion by cornhundred** [manzt/anywidget#296] [JavaScript] [answered]
Hi, I would like to try to use deck.gl in anywidget. Would you have any advice on how to go about doing this?

**Comment by cornhundred** [comment]
Would it be possible to modify this confetti example 

```
import anywidget

class ConfettiWidget(anywidget.AnyWidget):
    _esm = """
    import confetti from "https://esm.sh/canvas-confetti@1.6";

    export function render({ model, el }) {
      let btn = document.createElement("button");
      btn.classList.add("confetti-button");
      btn.innerHTML = "click me!";
      btn.addEventListener("click", () => {
        confetti();
      });
      el.appendChild(btn);
    }
    """
    _css="""
    .confetti-button { background-color: #ea580c; }
    .confetti-button:hover { background-color: #9a3412; }
    """

ConfettiWidget()
```

to use this url https://esm.sh/deck.gl@8.9.21

I'm experimenting using this [Colab notebook](https://colab.research.google.com/drive/1wThSD7uSbMTp_PHLM35iSRPqF2t-_MX9?usp=sharing)

**Comment by manzt** [comment, marked as answer]
I usually refrain from debugging JavaScript issues unrelated to **anywidget** itself. However, being a fan of deck.gl, I let my curiosity get the best of me.

Your problem seems to be importing deck.gl as a valid ESM via esm.sh. For whatever reason, deck.gl isn't packaged correctly to work with esm.sh out of the box. However, there are ways to parameterize esm.sh to control what is sent to the browser. I recommend familiarizing yourself with [their docs](https://esm.sh/#docs) if you plan to use this CDN.

I was able to resolve this issue by adding `?bundle` to have esm.sh pre-bundle all dependencies into a single JS file on the server (rather than relying in native imports).

```python
import anywidget
import traitlets

class DeckWidget(anywidget.AnyWidget):
    _esm = """
    import { Deck, ScatterplotLayer } from 'https://esm.sh/deck.gl@8.9.21?bundle';
    export function render({ model, el }) {
        let root = document.createElement("div");
        root.style.height = "300px";
        let deck = new Deck({
          parent: root,
          controller: true,
          initialViewState: { longitude: -122.45, latitude: 37.8, zoom: 15 },
          layers: [
            new ScatterplotLayer({
              data: [{ position: [-122.45, 37.8], color: [255, 0, 0], radius: 100}],
              getFillColor: d => d.color,
              getRadius: d => d.radius
            })
          ],
        });
        el.appendChild(root);
        return () => deck.finalize();
    }
    """
```


https://github.com/manzt/anywidget/assets/24403730/20b40d4e-a23c-4087-94b0-793707c46391



I'd like to keep these GitHub issues dedicated to surfacing bugs with **anywidget**, rather than for debugging custom JavaScript or dependencies. I've provided a detailed answer in this case with the hope that it can serve as a reference for others.

**Comment by cornhundred** [comment]
cool thanks!

**Comment by cornhundred** [comment]
@manzt , I'm also having trouble importing d3.js but I can bring that up on the d3 repo and link it here for reference.

https://github.com/d3/d3/discussions/3740

**Comment by manzt** [comment]
what version of d3? i have a section in the docs that imports d3 from esm.sh

**Comment by manzt** [comment]
https://anywidget.dev/blog/introducing-anywidget/#ecmascript-modules-to-the-rescue

**Comment by cornhundred** [comment]
Thanks @manzt, that worked!

```
import anywidget

class D3Widget(anywidget.AnyWidget):
    _esm = """

    import * as d3 from "https://esm.sh/d3@7";

    export function render({ model, el }) {

    console.log('something')
    console.log(d3)

    }
    """
    _css="""
    """

D3Widget()
```
